docs/developer/architecture/centralized-cli.rst

```diff
@@ -0,0 +1,57 @@
+Centralized CLI
+===============
+
+Context
+-------
+
+MetalK8s comes with a set of services to operate and monitor the K8s cluster.
+All operations that need to be performed by the Platform Administrator could be
+categorized as follow:
+- Cluster Resources Administration (Nodes, Volumes, Deployments, ...)
+- Cluster Administration (Install, Upgrade, Downgrade, Backup, Restore, ...)
+- Solution Administration (CRUD Environment, Import/Remove Solution, ...)
+- Cluster Service Administration (Configure Dex, Prometheus, Alert Manager, ...)
+
+K8s provides the kubctl CLI, which is distributed and enables all kind of
+interaction with all resources available in etcd DB but its usage often requires
+to build verbose yaml files. Also kubectl tool only expose the standard K8s
+resources but not the custom ones, such as Volumes.
+
+Currently, MetalK8s provides other set of CLI or manual procedures, but those
+are located in various locations, their usage may vary and they are not
+developed using the same technologies.
+
+This makes the CLI and associated documentation not super intuitive and it
+also makes the maintenance more expensive in the long term.
+
+The goal of the project is to provide MetalK8s administrator an intuitive and
+easy to use set of CLI in order to administrate and operate a finite set of
+functionalities.
+
+Because kubectl CLI is already in place and is well known by Kubernetes
+administrators, it will be used as a standard to follow for all other MetalK8s
+CLIs:
+- CLI follows kubctl style
+- CLI is not interactive
+- CLI should not require password input
+- CLI implementation relies on secured APIs
+- CLI support completion for easy discovery
```

Comment:
on medium to large platforms it becomes super slow if this feature supports autocompleting pods/svc/volumes names. this needs to be counted for

Reply:
Hmm for me we definitly not want completion for object name but just for different operation we can do with the CLI

Reply:
Hum, completion is shell-specific.
I'm not sure we want to thread into this direction.

I think it's easier and as important to have a useful `-h/--help`, like `foo -h` list the top level commands, then a `foo command -h` gives the help of `command` and so on.
Like what `kubectl` (and others) does.

Reply:
For me, autocompletion for both object names and operations is a nice to have. Btw, @sayf-eddine-scality what is the cause of the slowness when autocompleting pods/svc/volumes names??

Reply:
https://github.com/kubernetes/kubernetes/issues/82883
loading a huge number of objects and then parsing their metadata takes some time (counting also network latency) and this happens every time you press Tab.

Reply:
Auto completion for objects name is a bad idea, because to discover btw you need to access apiserver so you likely need a kubeconfig ... and yes if you have tones of objects ect ...

> I think it's easier and as important to have a useful -h/--help, like foo -h list the top level commands, then a foo command -h gives the help of command and so on.
Like what kubectl (and others) does.

Agree, personnaly think having nice help text is good enough but with `cmd` lib from python we can have also an interpreter with completion

Reply:
> Agree, personnaly think having nice help text is good enough but with cmd lib from python we can have also an interpreter with completion

Ha I was thinking about `argparse`, but if we have a lib that does it for us then let's have completion on subcommands :slightly_smiling_face:

Reply:
> Auto completion for objects name is a bad idea, because to discover btw you need to access apiserver so you likely need a kubeconfig ... and yes if you have tones of objects ect ...
> 
If we provide OIDC login creds at the very beginning of the script as you say then we can use the resulting OIDC auth token to generate a kubeconfig and use for subsequent operations

> > I think it's easier and as important to have a useful -h/--help, like foo -h list the top level commands, then a foo command -h gives the help of command and so on.
> > Like what kubectl (and others) does.
> 
> Agree, personnaly think having nice help text is good enough but with `cmd` lib from python we can have also an interpreter with completion